### DISCUSSION TITLE
InteractiveParser & Better Cache

### DISCUSSION THREAD

**Discussion by erezsh** [lark-parser/lark#885] [Releases]
## Cache
- Lark now tracks changes in imported grammars (`%import`), and updates the cache if necessary
- Added support for atomicwrites, for multiprocess caching and crash recovery

## InteractiveParser
- Now an official interface (renamed from Puppet)
- Added `Lark.parse_interactive()` for starting the parser in interactive mode

## Other

- Added `ast_utils`, to assist in tranforming `lark.Tree` into a customized AST.

- Better docs

- Bugfixes

## Notification: Support for Python 2 is ending

In the near future, Lark will drop support for Python 2. We will continue to develop for Python 3.6+ only, which will simplify the code and ease development.

Old releases (including this one) will still work, of course, and should be stable enough to accompany the remaining Python 2 users into the sunset.

If you have any objections, feel free to voice them here: https://github.com/lark-parser/lark/discussions/874

Thanks for everyone who helped make Lark better!

<hr /><em>This discussion was created from the release <a href='https://github.com/lark-parser/lark/releases/tag/0.11.3'>InteractiveParser & Better Cache</a>.</em>

**Comment by ThatXliner** [comment]
Just asking, are the docs updated for ast_utils? The current site for the docs doesn't include the [ast_utils example](https://github.com/lark-parser/lark/blob/master/examples/advanced/create_ast.py). Is it that the build for the documentation hasn't started or that readthedocs is super slow?

**Reply by erezsh** [reply]
It's just me who's super slow. It's there now: https://lark-parser.readthedocs.io/en/latest/classes.html#ast-utils

**Reply by ThatXliner** [reply]
> It's just me who's super slow.

Oh. I do remember readthedocs having an option for automatically building it.

> It's there now: https://lark-parser.readthedocs.io/en/latest/classes.html#ast-utils

Yay!

**Reply by ThatXliner** [reply]
Oh wait, [nevermind](https://github.com/lark-parser/lark/commit/6348538eb65207414a4bd7f014201306e84d2917)

**Reply by erezsh** [reply]
Yep.. In my defense, documentation is boring.
